Fix item permissions check

paescuj opened the pull request:
## Scope

What's changed:

Fixed the permission check in the Data Studio for permissions rules containing fields which the current user has no access to, via new dedicated permission check API endpoint.

## Potential Risks / Drawbacks

- Possibility of additional API request when opening an item in Data Studio
  - is implemented in such a way that it is only fetched when really necessary
- Ensure the new endpoint doesn't expose unwanted information
  - requires authentication
  - always returns the same structure, even if collection / item does not exist

## Review Notes / Questions

- New endpoint `/permissions/me/:collection/:id?`
  - The endpoint should only be called from within the Data Studio when necessary:
    - conditional permission rules (especially in combination with related collections, see #19327) since the Data Studio (client) might not have access to the corresponding fields
    -  singleton "update" permission since the Data Studio doesn't know whether to use "create" or "update" permission
- New SDK method `readItemPermissions(collection, id)`

---

- Fixes #19327
- Fixes issue where sort handler is displayed in tabular layout even if user has no update permission on sort field
- Fixes issue where update permission would be used for initial singleton creation, instead of create permission

paescuj committed:
Introduce permissions API endpoint to fix item permissions check

paescuj committed:
Add changeset

ComfortablyCoding commented:
@paescuj, LGTM so far!

Points of feedback for the current state:
- [ ]  We should throw a permission error for any collection/item the user/role does not have access to. 
     - At the moment it returns the same response whether or not the user/role has permissions and/or the collection/item exists. 
     - This paradigm of always returned a response is a different path than the one we currently use, it can lead to confusion. 
     - It is hard to determine what is considered "no access" in this case, might be worth using this new paradigm.
- [ ] We should be requesting the `/permissions/me` endpoint only when at least one of the access rules (update/delete/share) are not custom for the collection. 
     - Currently it is being requested each time an item is opened. 
     - It seems to also be doing a double request each time an item is opened.
     - This will cut down on requests made for collections where this is applicable.
- [ ] To be inline with other endpoints should we be wrapping this all in a data property?
- [ ] I do not believe we need to explicitly handle the admin case as [it is already done by checkAccess](https://github.com/directus/directus/blob/062c8f23f602f682aebcb0c015e43155d3ff6715/api/src/services/authorization.ts#L593).
    - This should result in it returning true for admin on all operations.
    - The performance impact should be minimal as it is early return. 

Non blocking UX improvement
- Mentioning this here but it is probably best left for another PR. Any relations in an item that a user/role does not have access to shows as configured incorrectly. This can lead to some confusion.

ComfortablyCoding reviewed (dismissed):
See the above feedback.

paescuj committed:
Revision

- Wrap API response in 'data' prop
- Split-up usePermissions composables & add tests
- Rework all permission checks

paescuj committed:
Merge branch 'main' into fix-item-permissions-check

paescuj committed:
Remove obsolete test

paescuj committed:
Merge branch 'main' into fix-item-permissions-check

paescuj committed:
Revision 2

- Singleton support
- Bug fixes
- Use permission composables wherever applicable
- Complete tests

paescuj committed:
Merge branch 'main' into fix-item-permissions-check

paescuj committed:
Update mock path in archive test

br41nslug reviewed (commented):
Not a complete review but so far looks good and fixes the issue in my tests 😄 just the small security idea for today

paescuj committed:
Move remaining checks to usePermissions composables

paescuj committed:
Update docs

paescuj committed:
Lint & spelling fix

paescuj committed:
Add SDK method

paescuj committed:
Merge branch 'main' into fix-item-permissions-check

paescuj commented on api/src/controllers/permissions.ts:
I agree! Obviously, this new endpoint is mainly intended for the Data Studio, where the user is authenticated anyway. Therefore, it makes sense to restrict the endpoint and not to take any unnecessary risks. We can still open it up at a later date if necessary, but for now I'm going to adjust it as suggested by you 👍

paescuj commented:
Thanks @ComfortablyCoding for the review ❤️ 

The points mentioned by you have been solved in the meantime 👍 

> We should throw a permission error for any collection/item the user/role does not have access to.
> 
> * At the moment it returns the same response whether or not the user/role has permissions and/or the collection/item exists.
> * This paradigm of always returned a response is a different path than the one we currently use, it can lead to confusion.
> * It is hard to determine what is considered "no access" in this case, might be worth using this new paradigm.

The endpoint behaves differently in this respect to prevent information about the existence of collections and items from being disclosed. The endpoint should only provide information on whether the permissions are granted or not. As requested in https://github.com/directus/directus/pull/21152#discussion_r1478633299 we'll now throw a Forbidden error if the user is unauthenticated.

Non blocking UX improvement

- Mentioning this here but it is probably best left for another PR. Any relations in an item that a user/role does not have access to shows as configured incorrectly. This can lead to some confusion.

I agree that a better message is required for such cases. Since it's indeed out of scope for this PR as it's not related to the same logic, it's now tracked in #21337.

paescuj committed:
Require authentication for getItemPermissions service

paescuj committed:
Add blackbox tests

ComfortablyCoding commented:
LGTM so far from a functionality standpoint! Seems to work as expected. Code review (if applicable) will follow.

ComfortablyCoding commented on api/src/services/authorization.ts:
```diff
@@ -606,9 +606,14 @@ export class AuthorizationService {
+		} else {
+			query.limit = 1;
+			const result = await itemsService.readByQuery(query, { permissionsAction: action });
+			const access = (action === 'create' && result) || result[0];
+			if (!access) throw new ForbiddenError();
```
What is the reasoning here for checking access like this? Is `result.length > 0` not enough?

On a side note `readByQuery` seems to always return an array. Empty arrays are truthy so we should specifically check the length.

paescuj replied:
For the `create` action we expect an empty array (because the singleton doesn't exist yet), while for the other actions `result[0]` is expected to defined.

The logic follows the other check above:
https://github.com/directus/directus/blob/2a2dc5fcd21057dd611c1594cbb39ddabb417d42/api/src/services/authorization.ts#L606-L608

We could however use something like this to be more explicit:

```js
const access = action === 'create' ? result.length === 0 : !!result[0];
if (!access) throw new ForbiddenError();
```

But still feels safer to do a `!result` check beforehand, in case `result` should not be defined for some reason (e.g. hooks 🤔)
```
if (!result) throw new ForbiddenError();
const access = action === 'create' ? result.length === 0 : !!result[0];
if (!access) throw new ForbiddenError();
```

paescuj replied:
Updated 👍

ComfortablyCoding commented on api/src/services/permissions.ts:
```diff
@@ -144,4 +145,89 @@ export class PermissionsService extends ItemsService {
+	async getItemPermissions(collection: string, primaryKey?: string): Promise<ItemPermissions> {
+		if (!this.accountability?.user) throw new ForbiddenError();
+
+		if (this.accountability?.admin)
+			return {
+				update: { access: true },
+				delete: { access: true },
+				share: { access: true },
+			};
```
```suggestion
		if (this.accountability?.admin) {
			return {
				update: { access: true },
				delete: { access: true },
				share: { access: true },
			};
		}
```

As this is technically multi line we should use brackets for consistency.

paescuj committed:
Use multi-line if

Co-authored-by: daedalus <44623501+ComfortablyCoding@users.noreply.github.com>

ComfortablyCoding commented on app/src/composables/use-permissions/collection/lib/is-action-allowed.ts:
```diff
@@ -0,0 +1,15 @@
+import { computed, unref } from 'vue';
+import { Collection } from '../../types';
+
+export const isActionAllowed = (collection: Collection, action: 'create' | 'read' | 'update' | 'delete') => {
```
```suggestion
export const isActionAllowed = (collection: Collection, action: 'create' | 'read' | 'update' | 'delete' | 'share') => {
```

paescuj replied:
Hmm I'd only list the actions that are actually used, `share` isn't one of those. If we need that later on we can still expand the list.

ComfortablyCoding replied:
Agreed, looks like I confused that with the item equivalent.

ComfortablyCoding commented on docs/reference/system/permissions.md:
````diff
@@ -820,3 +820,149 @@ const result = await client.request(deletePermissions(['56', '57']));
+```json
+{
+  "data": {
+     "update": {
+       "access": true,
+	   "presets": permission_presets,
+	   "fields": permission_fields
+    },
````
```suggestion
     "update": {
       "access": true,
	     "presets": permission_presets,
	     "fields": permission_fields
    },
```

ComfortablyCoding commented on docs/reference/system/permissions.md:
````diff
@@ -820,3 +820,149 @@ const result = await client.request(deletePermissions(['56', '57']));
+```json
+{
+  "data": {
+     "update": {
+       "access": true,
+	   "presets": {},
+	   "fields": ["*"]
+    },
````
```suggestion
     "update": {
       "access": true,
	     "presets": {},
	     "fields": ["*"]
    },
```

br41nslug commented on packages/composables/src/use-collection.ts:
```diff
@@ -1,17 +1,17 @@
-	sortField: ComputedRef<string | null>;
-	isSingleton: ComputedRef<boolean>;
-	accountabilityScope: ComputedRef<'all' | 'activity' | null>;
+	info: Ref<AppCollection | null>;
+	fields: Ref<Field[]>;
+	defaults: Ref<Record<string, any>>;
+	primaryKeyField: Ref<Field | null>;
+	userCreatedField: Ref<Field | null>;
+	sortField: Ref<string | null>;
+	isSingleton: Ref<boolean>;
+	accountabilityScope: Ref<'all' | 'activity' | null>;
```
Considering `useCollection` actually returns `ComputedRef`s should types not reflect this? Having these types as Ref may give the false impression that these values are writable instead of computed

paescuj replied:
Good point, thanks! Fixed now at other places too: 9c2a838

paescuj committed:
Format JSON code blocks

paescuj committed:
Use correct type for computed values

Co-authored-by: Brainslug <tim@brainslug.nl>

paescuj committed:
More explicit check & add comments

paescuj committed:
Merge branch 'main' into fix-item-permissions-check

ComfortablyCoding commented:
LGTM!, @br41nslug had some final testing he wanted to get to. I will leave it for him to approve/merge.

br41nslug reviewed (approved):
Works as expected in my tests and looks good 😄 happy to finally see those multiple `usePermissions` embedded in components cleaned up 🚀

paescuj commented on api/src/services/permissions.ts:
```diff
@@ -144,4 +145,90 @@ export class PermissionsService extends ItemsService {
 	}
+
+	async getItemPermissions(collection: string, primaryKey?: string): Promise<ItemPermissions> {
+		if (!this.accountability?.user) throw new ForbiddenError();
```
As discussed, we'll keep it like this for now, but we'll happily revisit if there's a demand for it from extension side 👍
